ISPN-13725 Use safe codec version for ping operations

fax4ever opened the pull request:
https://issues.redhat.com/browse/ISPN-14018

In order to keep to support the handshake with 3.x clients

fax4ever commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/UntracedHotRodOperation.java:
```diff
@@ -0,0 +1,29 @@
+import org.infinispan.client.hotrod.impl.transport.netty.ChannelFactory;
+
+@Immutable
+public abstract class UntracedHotRodOperation<T> extends HotRodOperation<T> {
+
+   protected UntracedHotRodOperation(short requestCode, short responseCode, Codec codec, int flags, Configuration cfg,
+                                     byte[] cacheName, AtomicInteger topologyId, ChannelFactory channelFactory,
+                                     DataFormat dataFormat) {
+      super(requestCode, responseCode, chooseCodec(codec), flags, cfg, cacheName, topologyId, channelFactory, dataFormat);
+   }
+
+   protected UntracedHotRodOperation(short requestCode, short responseCode, Codec codec, int flags, Configuration cfg,
+                                     byte[] cacheName, AtomicInteger topologyId, ChannelFactory channelFactory) {
+      super(requestCode, responseCode, chooseCodec(codec), flags, cfg, cacheName, topologyId, channelFactory);
+   }
+
+   private static Codec chooseCodec(Codec codec) {
+      return codec.supportsTracing() ? ProtocolVersion.getWithoutTracing() : codec;
+   }
+}
```
This is a very nice solution that I think it was alternative to adding the `safePingVersionCodec` field to the `OperationsFactory`.
I like both the solutions.
Let's decide which one we want to keep it

jabolina replied:
I did this as a way to be more explicit if the command should be traced or not. If we have more commands in the future that we do not want to trace we could just change from which class it extends. The version only using the negotiated codec is simpler though, no problems if we want to stick with that.

fax4ever replied:
It is a nice idea to mark the protocol as ready to support the tracing or not. But I think that the changes we've made with Hot Rod v.4.0 are not only for tracing, but have a more general purpose scope.
Moreover, in my opinion it looks too complex to me having a for loop to find the first safe Hot Rod codec version.
In general if something could be implemented with less code and a simpler solution, I prefer that way.

fax4ever commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/PingResponse.java:
```diff
@@ -126,7 +126,7 @@ void processResponse(Codec codec, ByteBuf buf, HeaderDecoder decoder) {
       PingResponse build(short status) {
          assert decoderState == 4 : "Invalid decoder state";
-         return new PingResponse(status, version, keyMediaType, valueMediaType, serverOps);
+         return new PingResponse(status, version.choose(serverVersion), keyMediaType, valueMediaType, serverOps);
```
I don't know it this could work the same:

```suggestion
ProtocolVersion handshakeVersion =
   // take min( maxClientSupportedVersion, maxServerSupportedVersion )
   (serverVersion == null || serverVersion.compareTo(version) >= 0) ?
   version : serverVersion;

return new PingResponse(status, handshakeVersion, keyMediaType, valueMediaType, serverOps);
```

jabolina replied:
The choose operation I've tried to always use the highest as possible. For a client `C` and server `S`:

* Both can trace -> the maximum version
* Neither can trace -> the maximum version
* `C` can trace and `S` can not -> `S`' version
* `S` can trace and `C` can not -> `C`'s version

I think that originally we were always using the most recent, so I was unsure to fallback only to the oldest.

fax4ever reviewed (commented):
Thank you @jabolina this this my two cents.
Your turn to review my changes here. Thanks

jabolina commented on server/tests/src/test/java/org/infinispan/server/functional/VersionsInteropTest.java:
```diff
@@ -0,0 +1,29 @@
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+@Category(Unstable.class)
```
I was not sure if we want to always test this, so I marked as unstable.

jabolina commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/OperationsFactory.java:
```diff
@@ -56,6 +58,8 @@ public class OperationsFactory implements HotRodConstants {
 
    private Codec codec;
 
+   private Codec safePingVersionCodec;
```
I think we can rename these two, something in the lines of configuredCodec and negotiatedCodec. Only now I realized that this class was supposed to be immutable, but I am not sure if we need to add something for visibility on the negotiated and make the codec final.

fax4ever replied:
The codec itself was not immutable before this contribution, the idea was to update the codec when the negotiated version. I think

jabolina commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/OperationsFactory.java:
```diff
@@ -75,7 +79,13 @@ public OperationsFactory(ChannelFactory channelFactory, String cacheName, boolea
+      this.codec = (negotiatedCodec != null) ? negotiatedCodec : codec;
+
+      // in order to support the handshake with 3.x clients
+      this.safePingVersionCodec = (codec instanceof Codec40) ? new Codec31() : codec;
```
so we dont need to rely on the instanceof:

```suggestion
      this.safePingVersionCodec = codec.supportsTracing() ? ProtocolVersion.getWithoutTracing() : codec;
```

fax4ever replied:
I would avoid a for loop to choose a version that we know already is the 3.1

jabolina commented:
CI is not happy with my last commit. I'll trigger it again tomorrow.

fax4ever commented:
@jabolina I'm not sure it depends on your commit. I'm going to restart the CI. Let's try ;)

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/ProtocolVersion.java:
```diff
@@ -92,4 +92,31 @@ public static ProtocolVersion getBestVersion(int version) {
       throw new IllegalArgumentException("Illegal version " + version);
    }
+
+   public static Codec getWithoutTracing() {
+      // We skip the last version (auto)
+      for (int i = VERSIONS.length - 2; i >= 0; i--) {
+         if (!VERSIONS[i].codec.supportsTracing())
+            return VERSIONS[i].codec;
+      }
+      throw new IllegalStateException("Codec without tracing not available!");
+   }
```
can be replaced with
`public static final ProtocolVersion NO_TRACING_VERSION = PROTOCOL_VERSION_31;`

fax4ever replied:
+1 from me, since we can avoid the loop, maybe the name could be something like: `SAFE_HANDSHAKE_PROTOCOL_VERSION`. I would avoid using the word `tracing`, since the change on Hot Rod codec has a wider scope.

tristantarrant replied:
I like `SAFE_HANDSHAKE_PROTOCOL_VERSION`

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/OperationsFactory.java:
```diff
@@ -56,6 +58,8 @@ public class OperationsFactory implements HotRodConstants {
 
    private Codec codec;
 
+   private Codec safePingVersionCodec;
```
if the code is overwritten by `UntracedHotRodOperation`, why do we need this field?

pruivo reviewed (changes requested):
Changes in `org.infinispan.server.test.junit4.InfinispanRuleBasicTest` done by the OpenTelemetry PR needs to be reverted.

fax4ever commented:
Thanks @pruivo for the review. After restarting the CI, it looks good to me.

tristantarrant commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/UntracedHotRodOperation.java:
```diff
@@ -0,0 +1,36 @@
+ * @param <T>
+ */
+@Immutable
+public abstract class UntracedHotRodOperation<T> extends HotRodOperation<T> {
```
following the above reasoning, I'd rename this class to `NeutralVersionHotRodOperation` or some such

fax4ever commented:
I removed the first two commits, since the solution implemented there was alternative to the one implemented by @jabolina.
We decided to take his solution, probably more OO style.
I applied all the changes @jabolina passed to me and squashed his commits.
We finally decided to use the "min of max client version and max server version" as the resolution algorithm for the version to use.

If CI works, I would merge it.
@tristantarrant @pruivo let us know if it is fine for you, thank you

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelFactory.java:
```diff
@@ -150,6 +150,10 @@ public void start(Codec codec, Configuration configuration, Marshaller marshalle
       pingServersIgnoreException();
    }
 
+   public Codec getNegotiatedCodec() {
```
I don't understand why this is required. can you explain?

fax4ever replied:
Sure. Basically more `OperationsFactory` s could be created from the same `ChannelFactory`. I don't know why this is possible, see for instance:

https://github.com/infinispan/infinispan/blob/f415ee6db418375bd7a2a309603357c4a7ce345d/client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCacheManager.java#L626

The handshake happens at `ChannelFactory` level, so if the client and the server decided to agree on a given version, all `OperationsFactory` s using this `ChannelFactory` should use the same version as well.
Otherwise, there will be some errors.

That's the reason for the commit https://github.com/infinispan/infinispan/pull/10228/commits/836f6b6d2433bdc08b65ab029a0f9eed97c0c7d4

pruivo replied:
still does not make sense to me. `ChannelFactory` does not interact with the `codec` and the `PingResponse` sets the `codec` after the handshake.

fax4ever replied:
The handshake sets the code to the `OperationsFactory` that is contained in the `ChannelFactory`, so from that moment any other `OperationsFactory` created from the same `ChannelFactory` should use the same possible updated codec version. If we don't do that and we create an ad admin operation after the handshake, using a v4. client and a v.3.1 server, we get errors, since the client would use the 4.0 not supported by the server.
We tested it with @jabolina yesterday evening.

jabolina replied:
@pruivo Every time `administration()` is called it instantiates a new factory. The new factory uses the configured codec, which may be using a different version that the one negotiated by the ping. This causes the `AdminOperation` to fail. We took this approach so we can use the negotiated version in other places. 

TBH I think we could completely remove the codec placed in `RemoteCacheManager` and retrieve the correct version from the factory if needed, but this could be in another  PR.

fax4ever replied:
> TBH I think we could completely remove the codec placed in `RemoteCacheManager` and retrieve the correct version from the factory if needed, but this could be in another PR.

I think the same, but maybe it is call for the next time ;)

pruivo replied:
it does not make any sense to create a new `OperationsFactory` every time `administration()` is invoked! Since you are changing the `ChannelFactory`, it is better to return the ` OperationsFactory` instead of the `Codec`

fax4ever replied:
The creation of a new `OperationFactory` every time `administration()` is invoked is something that was already present on the main branch. We didn't change that since I don't know this part of the code, maybe there are some security reasons, e.g. the admin operation can be performed by a special role user admin. @pruivo If we say that is safe, we could avoid to recreate the `OperationFactory` if the `ChannelFactory` already has one initialized. But it looks like a big change, for that reason we didn't do that. Let us know about it.

fax4ever replied:
Or (I was thinking) the new `OperationFactory` created from `administration()` call is supposed to be passed to another thread, that could be another reason for this instance creation, and the `OperationFactory` and I don't know if the instance is thread safe. But as I said I made the choice that had the minimal impact, since I don't know the code very well here.

pruivo replied:
Looking at the code, I don't see a reason to not reuse the `OperationFactory`
* security/authentication -> performed by `AuthHandler` using the credentials from `Configuration`
* thread safety -> all fields are final, no issue here.

fax4ever replied:
> thread safety -> all fields are final, no issue here.

Maybe some nested fields are not.

Anyway I'm going to add a commit to avoid any new `OperationFactory` instance creation from a `ChannelFactory` if the `ChannelFactory` already has a `OperationFactory` instance.
If you say it is safe, it is OK.
Thanks

fax4ever replied:
I realized that the are other cases where an operations factory is created using an already started client factory, cases where the cache name needs to be passed, since it will be used by the new operations factory.
For these further cases so it is not possible to use the cached operations factory on client factory.
Thus, in order to use the right (negotiated) codec version I think that we cannot remove these changes.

[full discussion here](https://infinispan.zulipchat.com/#narrow/stream/118645-infinispan/topic/operations.20factory.20.2F.20channel.20factory/near/290771627)

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/OperationsFactory.java:
```diff
@@ -75,7 +75,9 @@ public OperationsFactory(ChannelFactory channelFactory, String cacheName, boolea
       this.forceReturnValue = forceReturnValue;
-      this.codec = codec;
+
+      Codec negotiatedCodec = (channelFactory != null) ? channelFactory.getNegotiatedCodec() : null;
```
same here

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/NeutralVersionHotRodOperation.java:
```diff
@@ -0,0 +1,28 @@
+   }
+
+   private static Codec chooseCodec(Codec codec) {
+      return codec.isSafeWithLatest() ? ProtocolVersion.SAFE_HANDSHAKE_PROTOCOL_VERSION.getCodec() : codec;
```
shouldn't be the other way around?
```suggestion
      return codec.isSafeWithLatest() ? codec : ProtocolVersion.SAFE_HANDSHAKE_PROTOCOL_VERSION.getCodec();
```

fax4ever replied:
If the protocol is the 40 that is not safe for the handshake returns `true` and a safe protocol (e.g. 3.1.) will be used.
So I agree with Pedro that probably we could use another name for this method like `isUnsafeForTheHandshake", it would be more in line with the `SAFE_HANDSHAKE_PROTOCOL_VERSION` constant name.

fax4ever commented:
With the last push I changed the name of the method and a little also its JavaDoc.
Thanks again for the review @pruivo

pruivo commented:
@fax4ever don't forget the revert the changes in `org.infinispan.server.test.junit4.InfinispanRuleBasicTest` made by the open telemetry PR.

pruivo commented:
merged! thanks @fax4ever !
I'll do a follow-up PR later with the improvements.

fax4ever commented:
Thanks @pruivo

fax4ever commented:
> @fax4ever don't forget the revert the changes in `org.infinispan.server.test.junit4.InfinispanRuleBasicTest` made by the open telemetry PR.

Opened https://github.com/infinispan/infinispan/pull/10238
